### DISCUSSION TITLE
Install private NPM dependencies in post-upgrade task

### DISCUSSION THREAD

**Discussion by subesokun** [renovatebot/renovate#14446] [Request Help]
Hi,

I want to run a post-upgrade task for migrating my Angular dependencies. Unfortunately, I've a private package in my `package.json` and hence the `npm ci` command is always failing. When adding a `.npmrc` with the credentials then everything works but of course I don't want to have the credentials unencrypted within Git. I've added the private registry to the Renovate hosts rule file but it seems it has no effect. First after committing my `.npmrc` with the credentials it worked. Any idea what I'm missing or how I can debug this?

Here is the post-upgrade task I want to run

```json
{
  "packageRules": [
    {
      "matchPackageNames": ["@angular/core"],
      "postUpgradeTasks": {
        "commands": [
          "npm ci --ignore-scripts",
          "npx ng update {{{depName}}} --from={{{currentVersion}}} --to={{{newVersion}}} --migrate-only --allow-dirty --force"
        ],
        "fileFilters": ["**/**"]
      }
    }
  ]
}
```

and my config looks similar to

```json
{
  "hostRules": [
    {
      "matchHost": "https://myprivatereg/npm/",
      "hostType": "npm",
      "encrypted": {
        "token": "<Encrypted PAT Token>"
      }
    }
  ]
}
```

and my `.npmrc` (commited into the repo)

```
@myprivatereg:registry=https://myprivatereg/npm/
```

and here some debug logs

```
"hostRules": [
           {
             "matchHost": "https://myprivatereg/npm/",
             "hostType": "npm",
             "encrypted": {"token": "***********"}
           }
         ],
         "packageRules": [
           {
             "matchDataSources": ["npm"],
             "registryUrls": [
               "https://myprivatereg/npm/"
             ]
           }
         ]
```

Thanks!

**Comment by rarkins** [comment]
Exactly how are you running Renovate?

And do the hostRules you set work normally when private packages are being looked up or updated?

**Reply by subesokun** [reply]
We're running currently the WhiteSource for self-hosted GitLab setup. So Renovate is running as part of the wss remediate container. And yes, packages updates of our private packages are getting detected by Renovate. So general setup should be fine.

**Reply by rarkins** [reply]
So in this case, commands are all run within the same container.

If lookups as well as `npm install` works for you then it implies that Renovate is writing your `hostRules` to `.npmrc` correctly prior to calling `npm install`. If your post upgrade task is not working, it implies either that we're reverting the `.npmrc` file prior to the post upgrade task, or that it's being run in a different directory where it doesn't see the `.npmrc`.

Have you tried setting [skipInstalls](https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls) to `false` in your global config? This would mean that Renovate runs a "full" install instead of package lock only, which could also mean that the `node_modules` are already in place for you to run your post upgrade task in.

**Reply by subesokun** [reply]
I'll give [skipInstalls](https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls) a try. I've also executed `npm config list` as part of the post upgrade task but it had just the content of my `.npmrc`. So it seems like the `.npmrc` that has been used during the previous `npm install` gets reverted before the post upgrade tasks run.

**Reply by rarkins** [reply]
Yeah, I think that reverting is intentional, although I'm not certain it's *necessary*

**Reply by subesokun** [reply]
Running a full install really slows down the scanning process noticeably in our case hence it would be great if there is a way to keep the `.npmrc`.

**Reply by GKersten** [reply]
@subesokun were you able to find a solution for this? I seem to run into a similar issue at the moment.

**Reply by subesokun** [reply]
@GKersten Sorry, can't remember if we found a solution to that issue
